api/app/src/main/kotlin/packit/controllers/PacketController.kt

```diff
@@ -77,4 +81,29 @@ class PacketController(
             .headers(response.second)
             .body(response.first)
     }
+
+    @GetMapping("/{name}/{id}/zip")
+    @PreAuthorize("@authz.canReadPacket(#root, #id, #name)")
+    // TODO: Verify authorization for accessing specific files: whether user has access to all files, or artefacts only.
+    fun downloadZip(
+        @PathVariable id: String,
+        @PathVariable name: String,
+        @RequestParam(required = true) hashes: List<String>,
+        @RequestParam(required = true) filenames: List<String>,
+    ): ResponseEntity<StreamingResponseBody>
+    {
+        val filesWithNames = hashes.mapIndexed { index, hash ->
+            val fileResource = packetService.getFileByHash(hash, false, hash).first
```

Comment:
The call to `getFileByHash` is going to return the file contents as a `ByteArray`.

The `filesWithNames` variable is going to end up holding the entire packet in memory, which defeats all the work done to stream the output of the Zip.

What you need to do is extend the `PacketService` class to have a streaming version of `getFileByHash`. You need to move the calls to that function inside the `StreamingResponseBody` to make sure they happen on the thread pool, and stream the file bodies into the Zip without loading the whole contents in memory.